src/common/checksum.c

```diff
@@ -1191,6 +1191,142 @@ ci_buf2uint64(const uint8_t *buf, uint16_t len)
 	return 0;
 }
 
+static int
+dcs_csum_info_list_resize(struct dcs_ci_list *list, daos_size_t number_of_bytes_needed)
+{
+	uint8_t			*new_allocation;
+	uint32_t		 i;
+	uint32_t		 new_buf_size;
+
+	new_buf_size = (list->dcl_buf_size + number_of_bytes_needed) * 2;
+	D_REALLOC(new_allocation, list->dcl_csum_infos, list->dcl_buf_size, new_buf_size);
+
+	if (new_allocation == NULL)
+		return -DER_NOMEM;
+	list->dcl_csum_infos = new_allocation;
+	list->dcl_buf_size = new_buf_size;
+
+	/* need to rewire the csum buffers */
+	for (i = 0; i < list->dcl_csum_infos_nr; i++) {
+		struct dcs_csum_info *dst = dcs_csum_info_get(list, i);
+
+		dst->cs_csum = ((uint8_t *)dst) + sizeof(struct dcs_csum_info);
+	}
+
+	return 0;
+}
+
+static inline bool
+list_has_enough_space(struct dcs_ci_list *list, struct dcs_csum_info *info,
+		      daos_size_t *size_needed)
+{
+	*size_needed = sizeof(struct dcs_csum_info) + ci_csums_len(*info);
+
+	return list->dcl_buf_used + *size_needed <= list->dcl_buf_size;
+}
+
+static inline void
+copy_csum_info(struct dcs_csum_info *dst, struct dcs_csum_info *src)
+{
+	/* copy everything, then will update the csum buffer */
+	*dst = *src;
+	/*
+	 * For a csum_info list, the csum will always be stored just after the csum_info.
+	 * It is assumed that enough memory has been allocated to perform this copy.
+	 */
+	dst->cs_csum = ((uint8_t *)dst) + sizeof(struct dcs_csum_info);
+	memcpy(dst->cs_csum, src->cs_csum, dst->cs_buf_len);
+}
+
+static struct dcs_csum_info *
+csum_info_get_local(struct dcs_ci_list *list, uint32_t idx, bool check_idx)
+{
+	struct dcs_csum_info	*result;
+	int			 i;
+
+	if (check_idx && idx >= list->dcl_csum_infos_nr)
+		return NULL;
+
+	result = (struct dcs_csum_info *) &list->dcl_csum_infos[0];
+	for (i = 0; i < idx; i++)
```

Comment:
I'm not sure how many csums an IO request usually have, it's better to use a pointer array (or a pointer to the last csum) to track the csums instead of iterating the csums repeatedly.

Reply:
I had a discussion with Niu about this. Going to leave this approach for now, but will be an area for improvement in the future.